The primary interest of bioinformatics is to understand biological data. This is
typically achieved through the development of biology-aware software tools.
Nanopore sequencing, which emerged in the 1980s, is one approach employed in
bioinformatics to determining the primary structure of biopolymers such as DNA
and RNA \cite{three-decades-nano}. It involves recording the ionic current as a
biopolymer passes through a nanoscale protein pore (or \textit{nanopore}) with a
voltage applied to its surrounding membrane \cite{Wang2021}. See Figure
\ref{fig:nano} for a pictorial overview of nanopore sequencing. The resulting
ionic current signal is then used to determine the nucleotide sequence of the
biopolymer.

Oxford Nanopore Technologies (ONT) is the leading producer of
nanopore sequencing machines and is used in most peer-reviewed studies involving
nanopore sequencing. Several ONT sequencing machines are commercially available
including (in order of throughput) the MinION, GridION and PromethION.
Each machine uses one or many flow cells to sequence DNA or RNA molecules. Each
\textit{flow cell} contains thousands of nanopores which are grouped into fours
with each group corresponding to its own channel. Each channel is assigned an
electrode in the flow cell's sensor chip which is controlled and measured by an
application-specific integrated circuit (ASIC) chip \cite{Wang2021}. The nanopores in a flow
cell deteriorate during their sequencing lifetime since they are protein-based.
For example, a PromethION flow cell can produce a yield of up to 290 GiB and can
be reused up to five times using an ONT Wash Kit \cite{nano-web}.
The signal recorded by such machines is digitised and recorded as a sequence of
16-bit signed integers. This sequence, hereafter referred to as \textit{nanopore
signal data}, is the focus of this thesis.

\input{plots/nano}
% TODO put footnote in right place
\footnotetext[1]{Figure \ref{fig:nano} was taken from Fig. 1 in \cite{Wang2021}.}
